src/decoder/plugins/UsfDecoderPlugin.cxx

```diff
@@ -0,0 +1,342 @@
+    const int psf_version_state = psf_load( path, &stdio_callbacks, psf_version_tags, usf_loader, &state, usf_info, &state, 0 );
+
+    if ( psf_version_tags != 0x21 || psf_version_state <= 0 ) {
+        LogWarning(usf_domain, "Error loading usf file");
+        free(state.emu);
+        return;
+    }
+
+    usf_set_compare( state.emu, state.enable_compare );
+    usf_set_fifo_full( state.emu, state.enable_fifo_full );
+
+    int32_t sample_rate;
+    usf_render( state.emu, 0, 0, &sample_rate );
+
+    /* initialize the MPD decoder */
+
+    const AudioFormat audio_format(sample_rate, SampleFormat::S16, USF_CHANNELS);
+    assert(audio_format.IsValid());
+
+    double track_length;
+    double fade_time;
+    double total_length;
+    track_lengths(tags.length, tags.fade, track_length, fade_time, total_length);
+
+    // Duration
+    decoder_initialized(decoder, audio_format, true, SongTime::FromS(total_length));
+
+    /* .. and play */
+    DecoderCommand cmd;
+    signed short buf[USF_BUFFER_SAMPLES];
+    do {
+
+        const char* result = usf_render(state.emu, buf, USF_BUFFER_FRAMES, 0);
+        if (result != 0) {
+            LogWarning(usf_domain, result);
+            break;
+        }
+
+        // Simple fading
+        if (fade_time > 0 && total_length >= 0 && decoder.timestamp > track_length) {
+            const double vol = std::max(0.0, 1.0 - ((decoder.timestamp - track_length) / fade_time));
+            for (unsigned int i = 0; i < USF_BUFFER_SAMPLES; i++) {
+                buf[i] *= vol;
+            }
+        }
+
+        cmd = decoder_data(decoder, nullptr, buf, sizeof(buf), 0);
+
+        // Stop song manually
+        if (total_length >= 0 && decoder.timestamp > total_length+2)
+            break;
+
+        if (cmd == DecoderCommand::SEEK) {
+            // If user seeks during the fade period. Disable fading and play forever.
+            // Hacky way to give user posibility to enable looping on the fly
+            if (decoder.timestamp > track_length) {
+                total_length = -1;
+            }
+            // Seek manually by restarting emulator and discarding samples.
+            const double target_time = decoder_seek_time(decoder).ToDoubleS();
+            usf_restart(state.emu);
+            const unsigned frames_to_throw = (unsigned) (sample_rate*target_time+0.5);
+            usf_render(state.emu, 0, frames_to_throw, 0);
+
+            // Time correction after seek. Decided by trial and error.
+            decoder_command_finished(decoder);
+            double newTime = (frames_to_throw/(double)sample_rate)-0.5;
+            decoder_timestamp(decoder, newTime);
+        }
+
+    } while (cmd != DecoderCommand::STOP);
+    usf_shutdown(state.emu);
+    free(state.emu);
+}
+
+static void
+set_tag_from_tag(const struct TagHandler &handler, void *handler_ctx, TagType type, std::string value)
+{
+    if (!value.empty()) tag_handler_invoke_tag(handler, handler_ctx, type, value.c_str());
```

Comment:
Would it be preferable to set the tag to the empty string in this case?
All tags are not always present in the input files, which is why they can be empty.

I take it I should also pass the value string by reference?